Kh/fix coverage on next project

KarimHmaissi opened the pull request:
Adding coverage in terminal and auto generated coverage report on running NX test command. Modifies the next generator to update the project.json config

cloudratha commented on packages/next/src/generators/init/utils/createCustomTestConfig.ts:
```diff
@@ -0,0 +1,43 @@
+            codeCoverage: true,
+        },
+        configurations: {
+            ci: {
```
I think we should only run codeCoverage in ci by default.

cloudratha commented on packages/next/src/generators/init/utils/createCustomTestConfig.ts:
```diff
@@ -0,0 +1,43 @@
+    };
+}
+
+export default async function addCustomTestConfig(
```
Naming consistency to filename

KarimHmaissi replied:
This is my bad. Will update

cloudratha commented on packages/next/src/generators/init/utils/createCustomTestConfig.ts:
```diff
@@ -0,0 +1,43 @@
+import { joinPathFragments, Tree, updateJson } from '@nrwl/devkit';
```
File naming should be kebab-case for consistency.

KarimHmaissi replied:
This is my bad. Will update

cloudratha commented on packages/next/src/generators/init/utils/createCustomTestConfig.ts:
```diff
@@ -0,0 +1,43 @@
+        json => {
+            const updatedJson = json;
+            updatedJson.targets = updatedJson.targets ?? {};
+            updatedJson.targets.test = getTestConfig(projectSourceRoot);
```
Should we merge the configs with the existing test executor options?

KarimHmaissi replied:
Thats a good shout. Guess only really need to merge the CI coverage config with whats there already. Will update

KarimHmaissi commented:
Added a utility to the core library to merge project.json target configs. Also updated requested file names and ensured coverage is reported only in CI. While this change will only effect the next plugin I believe it should go in as it will add the utilities and further work can be done to ensure all plugins have the required testing config.

cloudratha commented on packages/common/core/src/lib/project-config.ts:
```diff
@@ -0,0 +1,63 @@
+function merge(a: unknown, b: unknown): unknown {
+    return isObject(a) && isObject(b)
+        ? deepMerge(a as Record<string, unknown>, b as Record<string, unknown>)
+        : isObject(a) && !isObject(b)
```
This doesn't seem to handle deepMerging arrays? But rather prefers which ever is an object?

KarimHmaissi commented:
Updated the merge with a combination of an NPM package [https://github.com/TehShrike/deepmerge ]( https://github.com/TehShrike/deepmerge ) and a custom array merge algorithm. It will now deep merge complex objects with nested arrays. Added the ability to specify a key in an object so the algorithm will treat that object differently. For example {files: '.tsx'}, {files: '.tsx'} will be merged together but {files: '.jsx'} will not. This will handle the cases such as merging the ESLint config overrides
